### ISSUE TITLE
False negative in advanced map destructuring

### ISSUE THREAD

**Issue by DjebbZ** [borkdude/clj-kondo#673]
**version**

```
> clj-kondo --version
clj-kondo v2019.12.14
```

**platform**

Linux, native executable.

**editor**

Just the command-line is enough to reproduce the bug.

**problem**

When doing some advanced destructing, clj-kondo doesn't understand the code and reports unused binding and unresolved symbols. This is pure Clojure. See repro example below.

**repro**

``` clojure
(let [{a :a
        {b :b} (name a)} {:a :a
                          "a" {:b :b}}]
  (println "value of b" b))
;; => :b
```

```
> clojure lint-false-error.clj
value of b :b
```

```
> clj-kondo --lint lint-false-error.clj
lint-false-error.clj:1:8: warning: unused binding a
lint-false-error.clj:2:22: error: unresolved symbol a
linting took 7ms, errors: 1, warnings: 1
```

**expected behavior**

No error and no warning since this is completely valid Clojure that works.

[ Optional: if you are interested in doing a PR yourself, please leave a note. ]

I like to contribute to open-source, but not sure I have time/skills to do it.

**Comment by borkdude** [comment]
It seems the bound value of a is used in the "second" binding, but since it's an associative destructuring, the bindings aren't ordered maybe, so I'm puzzled if this should be used at all. I can't find any examples of this in the Clojure documentation. I'm going to get some confirmation from the #clojure channel on Slack for this. If this is officially supported by Clojure, I'm going to add support for it.

**Comment by borkdude** [comment]
An example where this doesn't seem to work:

```
(let [{a :a, b a, c b} {:a :x, :x :y}] c)
nil
```

**Comment by borkdude** [comment]
According to a Clojure core contributor and the author of clojure.tools.reader this is not supposed to work. Hence clj-kondo won't support it.

**Comment by DjebbZ** [comment]
LOL I discovered something weird then. Thanks again for being so fast to answer.

**Comment by luposlip** [comment]
@borkdude is that also why clj-kondo (v2020.07.29) marks this last occurrence of `a` (in the `:or` map) as `unresolved symbol a`, even though it works just fine?

```clojure
(fn [a & [{:keys [b] :or {b a}}]])
```

**Comment by borkdude** [comment]
@luposlip See #782 and #916. Since @puredanger said it was undefined behavior clj-kondo doesn't support it and maybe it should warn against it. On the other hand, the topic keeps coming up a lot, so maybe, if the core team can confirm this isn't ever going to change, clj-kondo could support it.

**Comment by puredanger** [comment]
no plans to change - this is undefined and you should not do that

**Comment by borkdude** [comment]
@puredanger So a warning like 'Using previous binding as default in :or is undefined behavior` would be in order? (#916)

**Comment by puredanger** [comment]
sounds good to me
